File: crates/bindings-csharp/BSATN.Runtime/BSATN/Uuid.cs
namespace SpacetimeDB;

using System.Buffers.Binary;
using System.Runtime.InteropServices;
using SpacetimeDB.BSATN;

/// <summary>
/// A universally unique identifier (UUID).
///
/// Generate `NIL`, random (`v4`), and time-ordered (`v7`) UUIDs.
/// </summary>
[StructLayout(LayoutKind.Sequential)]
public readonly record struct Uuid : IEquatable<Uuid>, IComparable, IComparable<Uuid>
{
    private readonly U128 value;

    public Uuid(U128 val) => value = val;

    /// <summary>
    /// The nil UUID (all bits set to zero).
    /// </summary>
    public static readonly Uuid NIL = new(new U128());

    /// <summary>
    /// The max UUID (all bits set to one).
    /// </summary>
    public static readonly Uuid MAX = new(new U128(ulong.MaxValue, ulong.MaxValue));

    /// <summary>
    /// Create a UUIDv4 from explicit random bytes.
    /// </summary>
    /// <remarks>
    /// This method assumes the provided bytes are already sufficiently random;
    /// it will only set the appropriate bits for the UUID version and variant.
    /// </remarks>
    /// <example>
    /// <code>
    /// var randomBytes = new byte[16];
    /// RandomNumberGenerator.Fill(randomBytes);
    /// var uuid = Uuid.FromRandomBytesV4(randomBytes);
    /// Console.WriteLine(uuid);
    /// // Output: 166a036f-90f3-714a-9731-c8814b70e326
    /// </code>
    /// </example>
    public static Uuid FromRandomBytesV4(ReadOnlySpan<byte> randomBytes)
    {
        if (randomBytes.Length != 16)
        {
            throw new ArgumentException("Must be 16 bytes", nameof(randomBytes));
        }

        Span<byte> bytes = stackalloc byte[16];
        randomBytes.CopyTo(bytes);
        bytes[6] = (byte)((bytes[6] & 0x0F) | 0x40); // version 4
        bytes[8] = (byte)((bytes[8] & 0x3F) | 0x80); // variant RFC 4122

        return new(U128.FromBytesBigEndian(bytes));
    }

    public enum UuidVersion
    {
        Nil,
        V4,
        V7,
        Max,
    }

    /// <summary>
    /// Create a UUIDv7 from a UNIX timestamp (milliseconds) and 10 random bytes.
    /// </summary>
    /// <remarks>
    /// This method will set the variant field within the counter bytes without attempting to shift
    /// the surrounding data. Callers using the counter as a monotonic value should be careful not to
    /// store significant data in the 2 least significant bits of the 3rd byte.
    /// </remarks>
    /// <example>
    /// <code>
    /// var millis = 1686000000000L;
    /// var randomBytes = new byte[10];
    /// var uuid = Uuid.FromUnixMillisV7(millis, randomBytes);
    /// Console.WriteLine(uuid);
    /// // Output: 6e8d8801-005c-0070-8000-000000000000
    /// </code>
    /// </example>
    public static Uuid FromUnixMillisV7(
        long millisSinceUnixEpoch,
        ReadOnlySpan<byte> counterRandomBytes
    )
    {
        if (counterRandomBytes.Length != 10)
        {
            throw new ArgumentException(
                "counterRandomBytes must be exactly 10 bytes",
                nameof(counterRandomBytes)
            );
        }
        // Translated from Rust `uuid`
        var millisHigh = (uint)((millisSinceUnixEpoch >> 16) & 0xFFFF_FFFF);
        var millisLow = (ushort)(millisSinceUnixEpoch & 0xFFFF);

        var counterRandomVersion = (ushort)(
            ((counterRandomBytes[1] | (counterRandomBytes[0] << 8)) & 0x0FFF) | (0x7 << 12)
        ); // version = 7

        Span<byte> d4 =
        [
            // Variant + sequence bytes
            (byte)((counterRandomBytes[2] & 0x3F) | 0x80), // variant: 0b10xxxxxx
            counterRandomBytes[3],
            counterRandomBytes[4],
            counterRandomBytes[5],
            counterRandomBytes[6],
            counterRandomBytes[7],
            counterRandomBytes[8],
            counterRandomBytes[9],
        ];

        // Now assemble UUID bytes exactly like Rust's from_fields
        Span<byte> bytes = stackalloc byte[16];

        // millis_high → bytes[0..4] (big endian)
        bytes[0] = (byte)((millisHigh >> 24) & 0xFF);
        bytes[1] = (byte)((millisHigh >> 16) & 0xFF);
        bytes[2] = (byte)((millisHigh >> 8) & 0xFF);
        bytes[3] = (byte)(millisHigh & 0xFF);

        // millis_low → bytes[4..6] (big endian)
        bytes[4] = (byte)((millisLow >> 8) & 0xFF);
        bytes[5] = (byte)(millisLow & 0xFF);

        // counter_random_version → bytes[6..8] (big endian)
        bytes[6] = (byte)((counterRandomVersion >> 8) & 0xFF);
        bytes[7] = (byte)(counterRandomVersion & 0xFF);

        // d4 → bytes[8..16]
        d4.CopyTo(bytes[8..]);

        return new Uuid(U128.FromBytesBigEndian(bytes));
    }

    /// <summary>
    /// Generate a UUIDv7 using a monotonic <see cref="ClockGenerator"/>.
    /// </summary>
    /// <remarks>
    /// This method will set the variant field within the counter bytes without attempting to shift
    /// the surrounding data. Callers using the counter as a monotonic value should be careful not to
    /// store significant data in the 2 least significant bits of the 3rd byte.
    /// </remarks>
    /// <example>
    /// <code>
    /// var clock = new ClockGenerator(new Timestamp(1686000000000L));
    /// var randomBytes = new byte[10];
    /// var uuid = Uuid.FromClockV7(clock, randomBytes);
    /// Console.WriteLine(uuid);
    /// // Output: 7e640000-8151-0070-8000-000000000000
    /// </code>
    /// </example>
    public static Uuid FromClockV7(ClockGenerator clock, ReadOnlySpan<byte> randomBytes)
    {
        var millis = clock.Tick().ToUnixMillis();
        return FromUnixMillisV7(millis, randomBytes);
    }

    /// <summary>
    /// Returns the <see cref="UuidVersion"/> of this UUID.
    ///
    /// Throws <see cref="InvalidOperationException"/> if the UUID version is unknown.
    /// </summary>
    public UuidVersion GetVersion()
    {
        var bytes = value.ToBytesBigEndian();
        // variant is stored in the 7th byte, in the high nibble.
        var variant = (bytes[6] >> 4) & 0x0F;

        return variant switch
        {
            4 => UuidVersion.V4,
            7 => UuidVersion.V7,

            _ => this == Uuid.NIL ? UuidVersion.Nil
            : this == Uuid.MAX ? UuidVersion.Max
            : throw new InvalidOperationException("Unknown UUID version"),
        };
    }

    private static void GuidToBigEndianBytes(ReadOnlySpan<byte> guidBytes, Span<byte> be)
    {
        // time_low (4 bytes) — little-endian → reverse
        be[0] = guidBytes[3];
        be[1] = guidBytes[2];
        be[2] = guidBytes[1];
        be[3] = guidBytes[0];

        // time_mid (2 bytes) — little-endian
        be[4] = guidBytes[5];
        be[5] = guidBytes[4];

        // time_hi_and_version (2 bytes) — little-endian
        be[6] = guidBytes[7];
        be[7] = guidBytes[6];

        // last 8 bytes already big-endian
        guidBytes[8..].CopyTo(be[8..]);
    }

    private static void BigEndianBytesToGuid(ReadOnlySpan<byte> be, Span<byte> guidBytes)
    {
        // Guid’s weird internal layout (mixed-endian)

        // time_low (4 bytes) — little-endian
        guidBytes[0] = be[3];
        guidBytes[1] = be[2];
        guidBytes[2] = be[1];
        guidBytes[3] = be[0];

        // time_mid (2 bytes) — little-endian
        guidBytes[4] = be[5];
        guidBytes[5] = be[4];

        // time_hi_and_version (2 bytes) — little-endian
        guidBytes[6] = be[7];
        guidBytes[7] = be[6];

        // last 8 bytes already big-endian
        be[8..].CopyTo(guidBytes[8..]);
    }

    public static U128 FromGuid(Guid guid)
    {
        Span<byte> bytes = stackalloc byte[16];
        guid.TryWriteBytes(bytes);

        // Always interpret Guid bytes as big-endian
        return U128.FromBytesBigEndian(bytes);
    }

    /// <summary>
    /// Converts this instance to a <see cref="Guid"/>, in `big-endian`.
    /// </summary>
    public Guid ToGuid()
    {
        // Guid is `mixed-endian`, so we need to fixup
        Span<byte> be = stackalloc byte[16];
        BinaryPrimitives.WriteUInt64BigEndian(be[..8], value.Upper);
        BinaryPrimitives.WriteUInt64BigEndian(be[8..], value.Lower);

        Span<byte> gb = stackalloc byte[16];
        BigEndianBytesToGuid(be, gb);

        return new Guid(gb);
    }

    /// <summary>
    /// Parses a UUID from its string representation.
    /// </summary>
    /// <example>
    /// <code>
    /// var s = "01888d6e-5c00-7000-8000-000000000000";
    /// var uuid = Uuid.Parse(s);
    /// Console.WriteLine(uuid.ToString() == s); // True
    /// </code>
    /// </example>
    public static Uuid Parse(string s)
    {
        // Guid is `mixed-endian`, so we need to fixup
        var guid = new Guid(s);
        Span<byte> gb = stackalloc byte[16];
        guid.TryWriteBytes(gb);

        Span<byte> be = stackalloc byte[16];
        GuidToBigEndianBytes(gb, be);

        return new Uuid(U128.FromBytesBigEndian(be));
    }

    public override readonly string ToString()
    {
        return ToGuid().ToString();
    }

    public readonly int CompareTo(Uuid other) => value.CompareTo(other.value);

    /// <inheritdoc cref="IComparable.CompareTo(object)" />
    public int CompareTo(object? value)
    {
        if (value is Uuid other)
        {
            return CompareTo(other);
        }
        else if (value is null)
        {
            return 1;
        }
        else
        {
            throw new ArgumentException("Argument must be a Uuid", nameof(value));
        }
    }

    public static bool operator <(Uuid l, Uuid r) => l.CompareTo(r) < 0;

    public static bool operator >(Uuid l, Uuid r) => l.CompareTo(r) > 0;

    public readonly partial struct BSATN : IReadWrite<Uuid>
    {
        public Uuid Read(BinaryReader reader) => new(new SpacetimeDB.BSATN.U128Stdb().Read(reader));

        public void Write(BinaryWriter writer, Uuid value) =>
            new SpacetimeDB.BSATN.U128Stdb().Write(writer, value.value);

        // --- / auto-generated ---

        // --- customized ---
        public AlgebraicType GetAlgebraicType(ITypeRegistrar registrar) =>
            // Return a Product directly, not a Ref, because this is a special type.
            new AlgebraicType.Product(
                [
                    // Using this specific name here is important.
                    new("__uuid__", new AlgebraicType.U128(default)),
                ]
            );
    }
}

File: crates/bindings-typescript/src/lib/uuid.ts
import { ClockGenerator } from './timestamp';
import { AlgebraicType } from './algebraic_type.ts';

export type UuidAlgebraicType = {
  tag: 'Product';
  value: {
    elements: [
      {
        name: '__uuid__';
        algebraicType: { tag: 'U128' };
      },
    ];
  };
};

type UuidVersion = 'Nil' | 'V4' | 'V7' | 'Max';

/** 128-bit UUID abstraction */
export class Uuid {
  __uuid__: bigint;

  /** The nil UUID (all zeros). */
  static readonly NIL = new Uuid(0n);
  static readonly MAX_UUID_BIGINT = 0xffffffffffffffffffffffffffffffffn;
  /** The maximum UUID (all ones). */
  static readonly MAX = new Uuid(Uuid.MAX_UUID_BIGINT);

  /** Create a UUID from a bigint (u128).
   *  @throws Error if the bigint is not between 0 and `MAX_UUID_BIGINT`.
   * */
  constructor(u: bigint) {
    // Must fit in exactly 16 bytes
    if (u < 0n || u > Uuid.MAX_UUID_BIGINT) {
      throw new Error('Invalid UUID: must be between 0 and `MAX_UUID_BIGINT`');
    }
    this.__uuid__ = u;
  }

  /** Create a UUIDv4 from explicit random bytes.
   * @throws Error if the byte array is not exactly 16 bytes.
   * */
  static fromRandomBytesV4(bytes: Uint8Array): Uuid {
    if (bytes.length !== 16) throw new Error('UUIDv4 requires 16 bytes');
    const arr = new Uint8Array(bytes);
    arr[6] = (arr[6] & 0x0f) | 0x40; // version 4
    arr[8] = (arr[8] & 0x3f) | 0x80; // variant
    return new Uuid(Uuid.bytesToBigInt(arr));
  }

  /** Create a UUIDv7 from a UNIX timestamp (milliseconds) and 10 random bytes.
   * @throws Error if the byte array is not exactly 10 bytes.
   * */
  static fromUnixMillisV7(
    millisSinceUnixEpoch: bigint,
    counterRandomBytes: Uint8Array
  ): Uuid {
    if (counterRandomBytes.length !== 10) {
      throw new Error('UUIDv7 requires 10 random bytes');
    }
    // Translated from Rust `uuid`
    const millis = millisSinceUnixEpoch;

    const millisHigh = Number((millis >> 16n) & 0xffff_ffffn);
    const millisLow = Number(millis & 0xffffn);

    const counterRandom =
      (((counterRandomBytes[0] << 8) | counterRandomBytes[1]) & 0x0fff) |
      (0x7 << 12);

    const bytes = new Uint8Array(16);

    bytes[0] = (millisHigh >>> 24) & 0xff;
    bytes[1] = (millisHigh >>> 16) & 0xff;
    bytes[2] = (millisHigh >>> 8) & 0xff;
    bytes[3] = millisHigh & 0xff;

    bytes[4] = (millisLow >>> 8) & 0xff;
    bytes[5] = millisLow & 0xff;

    bytes[6] = (counterRandom >>> 8) & 0xff;
    bytes[7] = counterRandom & 0xff;

    bytes[8] = (counterRandomBytes[2] & 0x3f) | 0x80;
    bytes[9] = counterRandomBytes[3];
    bytes[10] = counterRandomBytes[4];
    bytes[11] = counterRandomBytes[5];
    bytes[12] = counterRandomBytes[6];
    bytes[13] = counterRandomBytes[7];
    bytes[14] = counterRandomBytes[8];
    bytes[15] = counterRandomBytes[9];

    return new Uuid(Uuid.bytesToBigInt(bytes));
  }

  /** Generate a UUIDv7 using a monotonic clock generator.
   * @throws Error if the random byte array is not exactly 10 bytes.
   * */
  static fromClockV7(clock: ClockGenerator, random: Uint8Array): Uuid {
    const millis = clock.tick().toMillis();
    return Uuid.fromUnixMillisV7(millis, random);
  }

  /** Parse a UUID from a string.
   * @throws Error if the string is not a valid UUID format.
   * */
  static parse(s: string): Uuid {
    const hex = s.replace(/-/g, '');
    if (hex.length !== 32) throw new Error('Invalid hex UUID');

    let v = 0n;
    for (let i = 0; i < 32; i += 2) {
      v = (v << 8n) | BigInt(parseInt(hex.slice(i, i + 2), 16));
    }
    return new Uuid(v);
  }

  /** Convert to string (hyphenated form). */
  toString(): string {
    const bytes = Uuid.bigIntToBytes(this.__uuid__);
    const hex = [...bytes].map(b => b.toString(16).padStart(2, '0')).join('');

    // Format as 8-4-4-4-12
    return (
      hex.slice(0, 8) +
      '-' +
      hex.slice(8, 12) +
      '-' +
      hex.slice(12, 16) +
      '-' +
      hex.slice(16, 20) +
      '-' +
      hex.slice(20)
    );
  }

  /** Convert to bigint (u128). */
  asBigInt(): bigint {
    return this.__uuid__;
  }

  /** Return a `Uint8Array` of 16 bytes. */
  toBytes(): Uint8Array {
    return Uuid.bigIntToBytes(this.__uuid__);
  }

  private static bytesToBigInt(bytes: Uint8Array): bigint {
    let result = 0n;
    for (const b of bytes) result = (result << 8n) | BigInt(b);
    return result;
  }

  private static bigIntToBytes(value: bigint): Uint8Array {
    const bytes = new Uint8Array(16);
    for (let i = 15; i >= 0; i--) {
      bytes[i] = Number(value & 0xffn);
      value >>= 8n;
    }
    return bytes;
  }

  /**
   * Returns the version of this UUID.
   * @throws Error if the UUID version is unsupported.
   */
  getVersion(): UuidVersion {
    const version = (this.toBytes()[6] >> 4) & 0x0f;

    switch (version) {
      case 4:
        return 'V4';
      case 7:
        return 'V7';
      default:
        if (this == Uuid.NIL) {
          return 'Nil';
        }
        if (this == Uuid.MAX) {
          return 'Max';
        }
        throw new Error(`Unsupported UUID version: ${version}`);
    }
  }

  compareTo(other: Uuid): number {
    if (this.__uuid__ < other.__uuid__) return -1;
    if (this.__uuid__ > other.__uuid__) return 1;

    return 0;
  }

  static getAlgebraicType(): UuidAlgebraicType {
    return AlgebraicType.Product({
      elements: [
        {
          name: '__uuid__',
          algebraicType: AlgebraicType.U128,
        },
      ],
    });
  }
}
